contrib/babelfishpg_tsql/src/tsql_for/forjson.c

```diff
@@ -103,45 +136,28 @@ tsql_query_to_json_sfunc(PG_FUNCTION_ARGS)
 	}
 
 	MemoryContextSwitchTo(old_context);
-
 	PG_RETURN_POINTER(state);
 }
 
-PG_FUNCTION_INFO_V1(tsql_query_to_json_ffunc);
-
-Datum
-tsql_query_to_json_ffunc(PG_FUNCTION_ARGS)
+// Main row to json function. 
+// Creates a Jsonb row object, processes the row, determines if it should be inserted as a nested json object
+// inserts json object to row and then into the main jsonbArray.
+static void
+tsql_row_to_json(JsonbValue* jsonbArray, Datum record, bool include_null_values)
 {
-	StringInfo	res = makeStringInfo();
-	char	   *state = ((StringInfo) PG_GETARG_POINTER(0))->data;
+	// HashTable
+	HTAB	   *jsonbHash;
```

Comment:
Is it possible to have this hash table persist between invocations? Because I believe that the nested structure should be the same for each row. You could initialize it in the initfunc and have it stored in the state struct. This would eliminate unnecessarily recreating the hash table for each row.